packages/liveblocks-react/src/umbrella-store.ts

```diff
@@ -588,9 +604,225 @@ export type CleanNotifications = {
+  const signal = new MutableSignal<NotificationsLUT>(new Map());
+
+  function markRead(notificationId: string, readAt: Date) {
+    signal.mutate((lut) => {
+      const existing = lut.get(notificationId);
+      if (!existing) {
+        return false;
+      }
+      lut.set(notificationId, { ...existing, readAt });
+      return true;
+    });
+  }
+
+  function markAllRead(readAt: Date) {
+    return signal.mutate((lut) => {
+      for (const n of lut.values()) {
+        n.readAt = readAt;
+      }
+    });
+  }
+
+  function deleteOne(inboxNotificationId: string) {
+    signal.mutate((lut) => lut.delete(inboxNotificationId));
+  }
+
+  function clear() {
+    signal.mutate((lut) => lut.clear());
+  }
+
+  function applyDelta(
+    newInboxNotifications: InboxNotificationData[],
+    deletedNotifications: InboxNotificationDeleteInfo[]
+  ) {
+    signal.mutate((lut) => {
+      let mutated = false;
+
+      // Add new notifications or update existing notifications if the existing notification is older than the new notification.
+      for (const n of newInboxNotifications) {
+        const existing = lut.get(n.id);
+        // If the notification already exists, we need to compare the two notifications to determine which one is newer.
+        if (existing) {
+          const result = compareInboxNotifications(existing, n);
+          // If the existing notification is newer than the new notification, we do not update the existing notification.
+          if (result === 1) continue;
+        }
+
+        // If the new notification is newer than the existing notification, we update the existing notification.
+        lut.set(n.id, n);
+        mutated = true;
+      }
+
+      for (const n of deletedNotifications) {
+        lut.delete(n.id);
+        mutated = true;
+      }
+      return mutated;
+    });
+  }
+
+  function updateAssociatedNotification(newComment: CommentData) {
+    signal.mutate((lut) => {
+      const existing = find(
+        lut.values(),
+        (notification) =>
+          notification.kind === "thread" &&
+          notification.threadId === newComment.threadId
+      );
+      if (!existing) return false; // Nothing to udate here
+
+      // If the thread has an inbox notification associated with it, we update the notification's `notifiedAt` and `readAt` values
+      lut.set(existing.id, {
+        ...existing,
+        notifiedAt: newComment.createdAt,
+        readAt: newComment.createdAt,
+      });
+      return true;
+    });
+  }
+
+  return {
+    signal: signal.asReadonly(),
+
+    // Mutations
+    markAllRead,
+    markRead,
+    delete: deleteOne,
+    applyDelta,
+    clear,
+    updateAssociatedNotification,
+
+    // XXX_vincent Remove this eventually
+    force_set: (
+      mutationCallback: (lut: NotificationsLUT) => void | undefined | boolean
+    ) => signal.mutate(mutationCallback),
+    invalidate: () => signal.mutate(),
+  };
+}
+
+function createStore_forRoomNotificationSettings() {
```

Comment:
Can't we type the return type of `createStore_*` functions?
I'd like to ensure we always set `signal` and `invalidate` and `force_set` as optional such as:
```ts
type Store<T> = {
    signal: SignalAsReadonly,
    invalidate: () => void,
    force_set?: () => void
} & T;
function createStore_whatever(): Store<{...}> {}
```

Reply:
I will do it once the actual types have been settled. The `force_set` and `invalidate` methods should both go away soon (they're anti-patterns), but some code currently still relies on it.

🔜